MoveIt Planning Pipeline Refactoring (#842)
ros-planning/moveit.ros.org: _posts/2024-03-25-MoveIt-Planning-Pipeline-Refactoring.md

---
author: Sebastian Jahr, Robotics Engineer
comments: false
date: 2024-03-25 00:01
layout: post

> title: "MoveIt Planning Pipeline Refactoring"

Reviewer: ```suggestion
title: "MoveIt 2 Planning Pipeline Refactoring"
```

<p class="m-0">However, the planning pipeline implementation in MoveIt had some shortcomings:</p>
- Order of algorithm calls was determined by the algorithm implementations, which made it hard to comprehend and prone to errors.
- It was not possible to chain planner algorithms.
- No clear separation and naming of pre-processing, planning and post-processing algorithms.

This is why we decided to refactor the planning pipeline to address the problems above (tracked by [moveit2#2408](https://github.com/ros-planning/moveit2/issues/2408)). This blog post describes these changes in detail.

### Refactoring Steps

Our goal was to refactor the current pipeline implementation (caricatural control flow shown in Fig. 1)

<div style="text-align:center">
    <img style="margin: 10px auto 32px;" src="{{ site.url }}/assets/images/blog_posts/planing_pipeline/image4.png" alt="Control flow of the old planning pipeline implementation" />
    <p>Fig. 1: Control flow of the old planning pipeline implementation</p>
</div>

into a clean and tidy structure as shown in Fig. 2

<div style="text-align:center">

>     <img style="margin: 10px auto 32px;" src="{{ site.url }}/assets/images/blog_posts/planing_pipeline/image1.png" alt="Refactrored planning pipeline implementation" />

Reviewer 2: ```suggestion
    <img style="margin: 10px auto 32px;" src="{{ site.url }}/assets/images/blog_posts/planing_pipeline/image1.png" alt="Refactored planning pipeline implementation" />
```

possibly in more places.

>     <p>Fig. 2: Refactrored planning pipeline implementation</p>

Reviewer: ```suggestion
    <p>Fig. 2: Refactored planning pipeline implementation</p>
```

**2.Test coverage and introspection**

<p class="m-0">Second, we improved the testing and introspection capabilities of the planning pipeline.</p>
- Unit Tests for the planning pipeline class have been added 🤓.
- Intermediate results between stages of the planning pipeline can now be published if the progress_topic parameter is set.

**3.Enable planner chaining**

<p class="m-0">The final step was to enable chaining planner algorithms.</p>
- A planning pipeline can load multiple planner plugins
    - You can now create seed trajectories for optimizing planner plugins with other planner plugins! For example, use OMPL’s RRTConnect to create a solution and optimize it with STOMP.
    - STOMP and CHOMP planner plugins can now use the reference trajectory in the MotionPlanResponse if available. While warm-starting optimization-based planning algorithms is not a new feature, it is now possible to do so by chaining planner plugins instead of writing an additional smoothing plugin. The chaining approach enables leaner and more flexible planning pipeline configurations.
- Removes duplicate code

>     - Before:  Two implementations of STOMP/CHOMP plugins existed: A PlannerPlugin and a RequestAdapterPlugin.

Reviewer: ```suggestion
    - Before: Two implementations of STOMP/CHOMP plugins existed: A PlannerPlugin and a RequestAdapterPlugin.
```

PILZIndustrialMotionPlanner<br/>
&nbsp;&nbsp;&nbsp;Generates industrial Cartesian trajectories (LIN, PTP, CIRC). You can find more information on Cartesian planning [in another blog post on this website](https://picknik.ai/cartesian%20planners/moveit/motion%20planning/2021/01/07/guide-to-cartesian-planners-in-moveit.html).<br/>

**PlanningResponseAdapters**<br/>
*ValidateSolution*<br/>
&nbsp;&nbsp;&nbsp;Checks the validity (collision-free, feasible, constraint satisfaction) of the path created by the planners.<br/>
*AddTimeOptimalParameterization*<br/>
&nbsp;&nbsp;&nbsp;Generates a time-optimal trajectory based on a given path based on an algorithm described by [Agarwal&nbsp;et&nbsp;al.](https://ieeexplore.ieee.org/document/6577978).<br/>
*AddRuckigTrajectorySmoothing*<br/>
&nbsp;&nbsp;&nbsp;Uses the [Ruckig](https://ruckig.com/) algorithm to adapt a trajectory to be jerk-constrained and time-optimal.<br/>
*DisplayMotionPath*<br/>
&nbsp;&nbsp;&nbsp;Publishes the computed path to RVIZ for visualization.<br/>

### Conclusion

> MoveIt now supports re-configuring planning pipelines to support the varying planning needs of unstructured environments. We are looking forward to seeing your planner or adapter plugin implementations as contributions to the MoveIt ecosystem! For more production environment features, check out our [MoveIt Pro](https://picknik.ai/pro/) offering.

Reviewer: ```suggestion
MoveIt 2 now supports more customizable planning pipelines to support the varying planning needs of unstructured environments. We are looking forward to seeing your planner or adapter plugin implementations as contributions to the MoveIt ecosystem! For more production environment features, check out our [MoveIt Pro](https://picknik.ai/pro/) offering.
```
